pkg/controller/controller_instance.go

```diff
@@ -212,1168 +211,520 @@ func (c *controller) reconcileServiceInstanceKey(key string) error {
+			readyCond := newServiceInstanceReadyCondition(v1beta1.ConditionFalse, reason, msg)
+			failedCond := newServiceInstanceFailedCondition(v1beta1.ConditionTrue, reason, msg)
+			return c.processProvisionFailure(instance, readyCond, failedCond, true)
 		}
 
-		message := fmt.Sprintf(
-			`Error deprovisioning, %s at ClusterServiceBroker %q: %v`,
-			pretty.ClusterServiceClassName(serviceClass), brokerName, err,
-		)
-		if httpErr, ok := osb.IsHTTPError(err); ok {
-			message = fmt.Sprintf("Deprovision call failed; received error response from broker: %v", httpErr)
+		// All other errors should be retried, unless the
+		// reconciliation retry time limit has passed.
+		msg := fmt.Sprintf("The provision call failed and will be retried: Error communicating with broker for provisioning: %v", err)
+		readyCond := newServiceInstanceReadyCondition(v1beta1.ConditionFalse, reason, msg)
+
+		if c.reconciliationRetryDurationExceeded(instance.Status.OperationStartTime) {
+			msg := "Stopping reconciliation retries because too much time has elapsed"
+			failedCond := newServiceInstanceFailedCondition(v1beta1.ConditionTrue, errorReconciliationRetryTimeoutReason, msg)
+			return c.processProvisionFailure(instance, readyCond, failedCond, false)
 		}
 
-		return c.updateServiceInstanceReconciliationError(
-			instance,
-			errorDeprovisionCalledReason,
-			message,
-		)
+		return c.processServiceInstanceOperationError(instance, readyCond)
 	}
 
 	if response.Async {
-		return c.updateServiceInstanceDeprovisionAsyncResponse(instance, response)
+		return c.processProvisionAsyncResponse(instance, response)
 	}
 
-	return c.updateServiceInstanceDeprovisionSuccess(instance)
+	return c.processProvisionSuccess(instance, response.DashboardURL)
 }
 
-// isServiceInstanceFailed returns whether the instance has a failed condition with
-// status true.
-func isServiceInstanceFailed(instance *v1beta1.ServiceInstance) bool {
-	for _, condition := range instance.Status.Conditions {
-		if condition.Type == v1beta1.ServiceInstanceConditionFailed && condition.Status == v1beta1.ConditionTrue {
-			return true
-		}
-	}
-
-	return false
-}
-
-// reconcileServiceInstance is the control-loop for reconciling Instances. An
-// error is returned to indicate that the instance has not been fully
-// processed and should be resubmitted at a later time.
-func (c *controller) reconcileServiceInstance(instance *v1beta1.ServiceInstance) error {
-	if instance.Status.AsyncOpInProgress {
-		return c.pollServiceInstance(instance)
-	}
-
-	if instance.ObjectMeta.DeletionTimestamp != nil || instance.Status.OrphanMitigationInProgress {
-		return c.reconcileServiceInstanceDelete(instance)
-	}
+// reconcileServiceInstanceUpdate is responsible for handling updating the plan
+// or parameters of a service instance.
+func (c *controller) reconcileServiceInstanceUpdate(instance *v1beta1.ServiceInstance) error {
```

Comment:
I am not keen on having two entirely separate 100-line functions, `reconcileServiceInstanceAdd` and `reconcileServiceInstanceUpdate`, that have the same workflow with just a few differences sprinkled in around which functions are called to do some of the work. I readily admit that the abundance of condition logic in the old `reconcileServiceInstance` function detracted from the readability of the function. However, there must be some middle ground that will allow us to avoid duplicating the workflow while also reducing the amount of conditional logic. Can we supply a strategy to the function that will supply the correct functions to use--a provision strategy and an update strategy? For example, instead of calling `c.processProvisionSuccess` we would call `processRequestSuccess` on the strategy, which in turn would call `c.processProvisionSuccess` when using a provision strategy?

Reply:
I agree having the two functions with near-identical reconciliation logic is not ideal, though I think the state of the refactor as it is is worth pushing through, and then we can optimize that aspect after if we want. There's some trickiness involved with updates not setting the Failed condition, especially with the retry duration expiring because then what normally would be the Failed condition becomes the Ready condition for updates. And extra scaffolding would add more single-use boilerplate functions with a bunch of switch statements. And then at that point I'd want to consider whether we might as well apply the scaffolding to deletion as well, and what that'd look like.

Between duplicating the workflow vs. having a bunch of switches inlined vs. generic function scaffolding with switches inside them, I felt at this time the workflow dup got the biggest quick bang for the buck with regards to readability and expandability.